### ISSUE TITLE
CVE-2019-2196/CVE-2019-2198 from ASB 2019.11

### ISSUE THREAD

**Issue by MSe1969** [cm12-amami/discussion#61]
Open Issue from #59 :
These are the LineageOS 14.1 commits for CVE-2019-2196/CVE-2019-2198:

https://github.com/LineageOS/android_frameworks_base/commit/598188b4602bdb4f091d61b190d056842c3e5efc
https://github.com/LineageOS/android_frameworks_base/commit/a634fae44691ac0ecf69ef1948657d0ed7b4b0dd
https://github.com/LineageOS/android_packages_providers_DownloadProvider/commit/f113a2f1acb0401e68a48b240938352a6467e25c
https://github.com/LineageOS/android_packages_providers_TelephonyProvider/commit/f744d319613245167cccf88b3b5a11b589244914
https://github.com/LineageOS/android_packages_providers_TelephonyProvider/commit/069d70878ce7cf834fae3461d33967e277a2324a

To backport to cm-12.1, the following needs consideration:
a. "double-colon" operator in 2nd commit (similar to 'lambda' behavior) - syntax not applicable in Java-7
```
    private void enforceStrictGrammar(@Nullable String selection, @Nullable String groupBy,
            @Nullable String having, @Nullable String sortOrder, @Nullable String limit) {
        SQLiteTokenizer.tokenize(selection, SQLiteTokenizer.OPTION_NONE,
                this::enforceStrictGrammarWhereHaving);
        SQLiteTokenizer.tokenize(groupBy, SQLiteTokenizer.OPTION_NONE,
                this::enforceStrictGrammarGroupBy);
        SQLiteTokenizer.tokenize(having, SQLiteTokenizer.OPTION_NONE,
                this::enforceStrictGrammarWhereHaving);
        SQLiteTokenizer.tokenize(sortOrder, SQLiteTokenizer.OPTION_NONE,
                this::enforceStrictGrammarOrderBy);
        SQLiteTokenizer.tokenize(limit, SQLiteTokenizer.OPTION_NONE,
                this::enforceStrictGrammarLimit);
    }
```

b. Check code in DownloadProvider and TelephonyProvider

**Comment by foresto** [comment]
I mostly avoid java, so I might not be the person to solve this one, but I'd like to take a look.

Can someone create a branch with these commits applied? That would make it easier to experiment with potential fixes.

**Comment by foresto** [comment]
I think I have the first two patches solved. I'm using anonymous classes to replace the `::` method references, along with a minimal implementation of Java 8's `java.util.function.Consumer` interface. This preserves most of the upstream code, including its structure, which should make future patches fairly easy to adapt.

I usually avoid Java, so review by an expert would be welcome, but several hours of reading have convinced me that this is a good solution.

**Comment by foresto** [comment]
I just noticed that the last two CVE patches (TelephonyProvider) look related to the July 2019 CVE-2019-2117 patch that #2 says was never applied. Perhaps it's time to revisit that one?

**Comment by MSe1969** [comment]
Thanks for the patch - I'll compile and test now.

In regards to CVE-2019-2117 - my comment in #55 was in fact a similar story than this issue - I considered the main blocking point our inability to convert a "lambda" into Java 7 code. 
Do you think you could have a look?
This is the respective AOSP commit: https://android.googlesource.com/platform/packages/providers/TelephonyProvider/+/ea309b07ac2309671a4462c8b747f731b58c6d5b

In regards to the a.m. DownloadProvider commit being part of CVE-2019-2196/CVE-2019-2198:
When looking for some missing commits, I found a comment in #42 in regards to CVE-2018-9493 of October 2018 ASB: 
"partly implemented; (DownloadProvider [2] n/a, as this would only apply, if the D/L search was enabled)"
Indeed, in LineageOS 14.1, the search functionality in Downloads is enabled, whilst it is not available in cm-12.1 - and hence the SQL injection vulnerability addressed in this patch (for current Android releases) would simply not apply, as there is no user input possible to perform an SQL injection. Thus I propose to skip the DownloadProvider patch - would you agree?

**Comment by foresto** [comment]
> In regards to CVE-2019-2117 - my comment in #55 was in fact a similar story than this issue - I considered the main blocking point our inability to convert a "lambda" into Java 7 code.
> Do you think you could have a look?

I think I can, but I'm not sure when yet. Might be a few days before I have time. A Quick glance tells me that it might make sense to address that one before committing the patches for this one.

**Comment by MSe1969** [comment]
I have merged the pull request, but had to amend the commits afterwards to fix the authorship and dates for the two 1st commits, but of course left the authorship of you commit

**Comment by MSe1969** [comment]
I'll have a look on the Telephony stuff, to bring at least the backport to the 'cm-12.1-test' branch - most probybly in the state where it does not compile. but maybe I will succeed in applying, what you did for the other repo...

**Comment by foresto** [comment]
> I have merged the pull request, but had to amend the commits afterwards to fix the authorship and dates for the two 1st commits

Thanks. I don't know how to directly apply commits that are taken from a different repo, and preserve their original metadata. Is there a git trick for this?

**Comment by MSe1969** [comment]
If you apply the commit manually, you would simply do, for example:
`git commit --author="Hall Liu <hallliu@google.com>" --date="Tue Apr 23 13:47:45 2019 -0700"`

If you fetch/pull from a different repo remotely, you would simply do `git cherry-pick <commit-hash>`

**Comment by MSe1969** [comment]
Could you please have a look - I am not really sure about what I did...

**Comment by MSe1969** [comment]
I think I finally made it in regards to TelephonyProvider. Build compiles and seems to work.
Please test on your end as well (replace TelephonyProvider in your local manifest).
If you confirm the positive testing and agree, that we do not need to patch DownloadProvider (because Search is not available in LP - see above), we can update the official manifest and close this issue. Looking forward to your feedback

**Comment by foresto** [comment]
I opened a new issue #63 for CVE-2019-2117, to reduce the clutter in here.

**Comment by foresto** [comment]
> in LineageOS 14.1, the search functionality in Downloads is enabled, whilst it is not available in cm-12.1 - and hence the SQL injection vulnerability addressed in this patch (for current Android releases) would simply not apply, as there is no user input possible to perform an SQL injection. Thus I propose to skip the DownloadProvider patch - would you agree?

Do the DownloadProvider patches for CVE-2018-9493 and for this one fail to apply or compile in cm-12.1? It seems to me that applying them, if possible without much tweaking, would have the benefit of making our lives easier when future security patches come along.

**Comment by MSe1969** [comment]
Fair point - would you like to take care? If so, please push to tree cm-12.1-test, so we can test first

**Comment by MSe1969** [comment]
I'm currently working on the December ASB - I'll have a look as well (weekend maybe), so we can ship all together, but if you're working already on it, please feel free to push

**Comment by foresto** [comment]
I've been trying to backport CVE-2018-9493's DownloadProvider patch since yesterday. It has been a bit of a pain, but I think I finally figured it out. If it works, and if I still have time, I'll move on to CVE-2019-2196/CVE-2019-2198's DownloadProvider patch (the remaining one in this issue).

**Comment by foresto** [comment]
> in LineageOS 14.1, the search functionality in Downloads is enabled, whilst it is not available in cm-12.1 - and hence the SQL injection vulnerability addressed in this patch (for current Android releases) would simply not apply, as there is no user input possible to perform an SQL injection. Thus I propose to skip the DownloadProvider patch - would you agree?

I don't know enough about the issue to agree. Download search via user input may be unavailable in cm-12.1, but what about search via other means? For example, could a malicious app trigger a search, including SQL injection, without any user interaction at all? [The CVE](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2018-9493) says, "User interaction is not needed for exploitation." Unless we can prove otherwise, it seems to me that fixing this would be wise. (I commented on this in #42 as well.)

I now have [this version](/LineageOS/android_packages_providers_DownloadProvider/commit/d728a508d651060676f93b252f5c62927a3728a1) of the fix backported to cm-12.1, along with all the patches from this CVE. Will install it on my phone once the build is done, so I can see if anything looks obviously broken. Will push my DownloadProvider updates to cm-12.1-test soon.

**Comment by foresto** [comment]
Looks like I don't have write permission to LineageOS/android_packages_providers_DownloadProvider, so I can't push.

**Comment by MSe1969** [comment]
Of course not, neither do I on **LineageOS** - our repo is cm12-amami, so you need to push to the fork in cm12-amami (cm-12.1-test branch)!

**Comment by MSe1969** [comment]
Just in case: In your local copy of DownloadProvider, create a branch cm-12.1-test (if not yet done) and check it out and make sure your patches are in.
Then, execute the below commands:
```
git remote add cm12amami https://cm12-amami/android_packages_providers_DownloadProvider
git push cm12-amami cm-12.1-test
```
(If you use SSH keys, replace "https://" with "git@github.com:")

**Comment by foresto** [comment]
> Of course not, neither do I on LineageOS

That "LineageOS" was a copy/paste mistake in the comment. I already have the correct remote & tracking branch in my local repo.

    $ git remote -v
    github	git@github.com:cm12-amami/android_packages_providers_DownloadProvider.git (fetch)
    github	git@github.com:cm12-amami/android_packages_providers_DownloadProvider.git (push)

    $ git branch -vv
      cm-12.1       302257a1 [github/cm-12.1] Remove "public" download feature.
    * cm-12.1-test  8670627f [github/cm-12.1-test: ahead 2] RESTRICT AUTOMERGE Enable stricter SQLiteQueryB

    $ git status
    On branch cm-12.1-test
    Your branch is ahead of 'github/cm-12.1-test' by 2 commits.
      (use "git push" to publish your local commits)

    $ git push
    ERROR: Permission to cm12-amami/android_packages_providers_DownloadProvider.git denied to foresto.
    fatal: Could not read from remote repository.

    Please make sure you have the correct access rights
    and the repository exists.

**Comment by MSe1969** [comment]
OK, I see - please check now.

**Comment by foresto** [comment]
Pushing works now. Thanks.

Both DownloadProvider backports are now on the cm-12.1-test branch.

**Comment by MSe1969** [comment]
The Download provider patch works - I was able to perform downloads.
I suggest to close this issue, I'll merge into the stable branch.
